src/lib.rs

```diff
@@ -1360,6 +1361,236 @@ where
+            fn to_i64(&self) -> Option<i64> {
+                self.to_integer().to_i64()
+            }
+
+            fn to_i128(&self) -> Option<i128> {
+                self.to_integer().to_i128()
+            }
+
+            fn to_u64(&self) -> Option<u64> {
+                self.to_integer().to_u64()
+            }
+
+            fn to_u128(&self) -> Option<u128> {
+                self.to_integer().to_u128()
+            }
+
+            fn to_f64(&self) -> Option<f64> {
+                Some(ratio_to_f64(
+                    self.numer as i128,
+                    self.denom as i128
+                ))
+            }
+        }
+    )*)
+}
+
+#[cfg(not(feature = "bigint"))]
+to_primitive_64!(u64 i64);
+
+#[cfg(all(target_pointer_width = "64", not(feature = "bigint")))]
+to_primitive_64!(usize isize);
+
+#[cfg(feature = "bigint")]
+impl<T: Clone + Integer + ToPrimitive + ToBigInt> ToPrimitive for Ratio<T> {
+    fn to_i64(&self) -> Option<i64> {
+        self.to_integer().to_i64()
+    }
+
+    fn to_i128(&self) -> Option<i128> {
+        self.to_integer().to_i128()
+    }
+
+    fn to_u64(&self) -> Option<u64> {
+        self.to_integer().to_u64()
+    }
+
+    fn to_u128(&self) -> Option<u128> {
+        self.to_integer().to_u128()
+    }
+
+    fn to_f64(&self) -> Option<f64> {
+        match (self.numer.to_i64(), self.denom.to_i64()) {
+            (Some(numer), Some(denom)) => Some(ratio_to_f64(
+                <i128 as From<_>>::from(numer),
+                <i128 as From<_>>::from(denom),
+            )),
+            _ => {
+                let numer: BigInt = self.numer.to_bigint()?;
+                let denom: BigInt = self.denom.to_bigint()?;
+                Some(ratio_to_f64(numer, denom))
+            }
+        }
+    }
+}
+
+trait Bits {
+    fn bits(&self) -> usize;
+}
+
+#[cfg(feature = "bigint")]
+impl Bits for BigInt {
+    fn bits(&self) -> usize {
+        self.bits()
+    }
+}
+
+impl Bits for i128 {
+    fn bits(&self) -> usize {
+        (128 - self.wrapping_abs().leading_zeros()) as usize
+    }
+}
+
+/// Converts a ratio of `T` to an f64.
+///
+/// In addition to stated trait bounds, `T` must be able to hold numbers 56 bits larger than
+/// the largest of `numer` and `denom`. This is automatically true if `T` is `BigInt`.
+fn ratio_to_f64<T: Bits + Clone + Integer + Signed + ShlAssign<usize> + ToPrimitive>(
+    numer: T,
+    denom: T,
+) -> f64 {
+    // Constants to avoid an std dependency. Copied from the values in std::f64.
+    const F64_MANTISSA_DIGITS: u32 = 53;
+    const F64_MAX_EXP: i32 = 1024;
+    const F64_MIN_EXP: i32 = -1021;
+    const F64_INFINITY: f64 = 1.0_f64 / 0.0_f64;
+
+    #[cfg(feature = "std")]
+    fn f64_exp2(expt: i32) -> f64 {
+        (expt as f64).exp2()
+    }
+
+    #[cfg(not(feature = "std"))]
+    fn f64_exp2(expt: i32) -> f64 {
+        let positive_power = 2f64.powi(expt.abs());
+        if expt < 0 {
+            1f64 / positive_power
```

Comment:
> It seems like `powi` is available even without `std`,

Ah, yes, because `FloatCore` is in scope -- so you're still using the `num-traits` implementation. That's fine.

`powi` should do the right thing for negative exponents already, no need for `abs` or the reciprocal here.

Reply:
Ah cool, does that mean I can just use `powi` regardless, and it will use the intrinsic if I have std and the num-trait implementation otherwise? If yes I can even remove the conditional `f64_exp2` function.

Reply:
That would work, but I just tested and found that `exp2` is fastest -- and surprisingly `num-traits` is faster than the "real" `powi`. It may be that `num-traits` is not as good with ULP, but that won't matter for base 2 anyway.

```
test exp2     ... bench:      12,848 ns/iter (+/- 919)
test num_powi ... bench:      35,807 ns/iter (+/- 1,063)
test powi     ... bench:      43,327 ns/iter (+/- 3,279)
```

(these times are for testing all of `f64::MIN_EXP..=f64::MAX_EXP`)

Reply:
Oh, you already changed -- that's probably fine, I doubt this is a bottleneck.